src/main/java/com/google/sps/daos/RealtimeDao.java

```diff
@@ -0,0 +1,74 @@
+package com.google.sps.daos;
+
+import java.io.IOException;
+
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.DatabaseReference;
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.FirebaseDatabase;
+import com.google.firebase.database.ValueEventListener;
+import com.google.sps.data.Counter;
+import java.io.IOException;
+
+public class RealtimeDao implements CounterDao {
+  
+  private Counter counter;
+
+  private DatabaseReference realtimeDb;
+
+  public RealtimeDao() {
+
+    try {
+      FirebaseOptions options = new FirebaseOptions.Builder()
+          .setCredentials(GoogleCredentials.getApplicationDefault())
+          .setDatabaseUrl("https://quizzy-step-2020.firebaseio.com/")
+          .build();
+
+      // Initialize DB with options
+      FirebaseApp.initializeApp(options);
+
+      // Reference to the whole Realtime DB
+      final FirebaseDatabase database = FirebaseDatabase.getInstance();
+
+      // Reference to the "counter" object
+      realtimeDb = database.getReference("/users-counter/counter");
+
+      // Add a listener to the counter, executed when counter is changed
+      realtimeDb.addValueEventListener(new ValueEventListener() {
+        @Override
+        public void onDataChange(DataSnapshot dataSnapshot) {
+          counter = dataSnapshot.getValue(Counter.class);
+          System.out.println("Listener executed " + counter.getValue());
+        }
+      
+        @Override
+        public void onCancelled(DatabaseError databaseError) {
+          System.out.println("The read failed: " + databaseError.getCode());
+        }
+      });
+
+    } catch(IOException ie) {
+      ie.printStackTrace();
+    }
+    
+  }
+
+  @Override
+  public Counter getCounter() {
+    return counter;
```

Comment:
Should this be reading the current counter value from the db?

Reply:
Okay so this is a design question I have: the way it is now, because of the listener added at the top of the file, counter gets updated every time there is a change, so the idea is that returning counter should mean that that counter is the most up to date counter. However, this is not very good design right? It should query the DB every time it is asked to do so right? Also, we probably would be wasting CPU on updating the variable counter even though it is not consulted by the user?

Reply:
1) It looks like the `counter` in memory here is never updated. So the code in `doGet()` will always result in writing `1`

2) You should be updating the counter value inside a [transaction](https://firebase.google.com/docs/database/admin/save-data#section-transactions). In the (probable) case that two requests to update the counter are being handled at the same time, they will both write the same value to the db unless you do the counter read/write inside a transaction.

Reply:
Regarding point 1, doesn't it get updated in line 43?

Agree with point #2, will do!

Reply:
Oh I see. You still have the same problem with concurrent updates using stale data doing it that way.

Reply:
Gotcha, aside from that, do you think that there would be a point on keeping the `counter` object in memory or should I just query it and send it to the user every time it is requested?

Reply:
Yeah it should be read from the db every time. The db is the single source of truth.